mode/verilog/verilog.js

```diff
@@ -354,11 +360,189 @@ CodeMirror.defineMode("verilog", function(config, parserConfig) {
+        this.svxCurCtlFlowChar = "";
+        this.svxPrevCtlFlowChar = "";
+        this.svxPrevPrevCtlFlowChar = "";
+        this.vxCodeActive = thisMode.vxModeIsOn;
+        this.vxIndentRq = 0;
+      },
+      "token": function(stream, state, _thisMode) {
+        if (state.vxContext.vxCodeActive==true && stream.sol() && state.vxContext.svxCurCtlFlowChar != "") {
+          state.vxContext.svxPrevPrevCtlFlowChar = state.vxContext.svxPrevCtlFlowChar;
+          state.vxContext.svxPrevCtlFlowChar = state.vxContext.svxCurCtlFlowChar;
+          state.vxContext.svxCurCtlFlowChar  = "";
+        }
+      },
+      "svxGenIndent": function(stream, state, thisMode) {
+        var rtnIndent = -1, indentUnitRq=0, curIndent=stream.indentation();
+        switch (state.vxContext.svxCurCtlFlowChar) {
+        case "\\":
+          curIndent = 0;
+          break;
+        case "|":
+          if (state.vxContext.svxPrevPrevCtlFlowChar=="@") {
+            indentUnitRq=-2; //-2 new pipe rq after cur pipe
+            break;
+          }
+          if (thisMode.svxchScopePrefixs.propertyIsEnumerable(state.vxContext.svxPrevCtlFlowChar))
+            indentUnitRq=1; // +1 new scope
+          break;
+        case "m4":
+          if (state.vxContext.svxPrevPrevCtlFlowChar=="@") {
+            indentUnitRq=-2; //-2 new inst rq after  pipe
+            break;
+          }
+          if (thisMode.svxchScopePrefixs.propertyIsEnumerable(state.vxContext.svxPrevCtlFlowChar))
+            indentUnitRq=1; // +1 new scope
+          break;
+        case "@":
+          if (state.vxContext.svxPrevCtlFlowChar=="S") {
+            indentUnitRq=-1; // new pipe stage after stmts
+          }
+          if (state.vxContext.svxPrevCtlFlowChar=="|") {
+            indentUnitRq=1; // 1st pipe stage
+          }
+          break;
+        case "S":
+          if (state.vxContext.svxPrevCtlFlowChar=="@") {
+            indentUnitRq=1; // flow in pipe stage
+          }
+          if (thisMode.svxchScopePrefixs.propertyIsEnumerable(state.vxContext.svxPrevCtlFlowChar))
+            indentUnitRq=1; // +1 new scope
+          break;
+        }
+        var statementIndentUnit = thisMode.svxindentUnit;
+        rtnIndent = curIndent + (indentUnitRq*statementIndentUnit);
+        return( (rtnIndent>=0) ?rtnIndent :curIndent );
+      },
+      "indent": function(state) {
+        return( (state.vxContext.vxCodeActive==true) ? state.vxContext.vxIndentRq : -1);
+      }
+    },
+    svxindentUnit: 2,
+    svxkeywords: {
+      "\\SVX":"keyword", "\\SV":"keyword", "m4":"def", "M4":"def", "\\SVX_version":"keyword"},
+    svxkpScopePrefixs: {
+      "**":"variable-2", "*":"variable-2", "$$":"variable", "$":"variable",
+      "^^":"attribute", "^":"attribute"},
+    svxchScopePrefixs: {
+      ">":"property", "->":"property", "-":"hr", "|":"link", "?$":"qualifier", "?*":"qualifier",
+      "@-":"variable-3", "@":"variable-3", "?":"qualifier"},
+    svxisOperatorChar:  /[\[\]=:]/
```

Comment:
The 5 properties above could simply be variables. That'd also remove the need to pass the `thisMode` around, I believe.